### PR TITLE
feat(tui): add quit confirmation modal before exiting

### LINKED ISSUE #273: Add a confirmation modal before exiting
A simple confirmation modal when exiting.

Image

> the above is a screenshot of the modal shown by [crush](https://github.com/charmbracelet/crush) (for idea only)

### PR DESCRIPTION
**Pull request by standard-librarian**
Closes #273

## What this does

When the user presses `ctrl+c` or `ctrl+q`, instead of shutting down immediately, a confirmation modal appears with two navigable buttons.

## How it works

- `←` / `→` (or `h`/`l`, `tab`/`shift+tab`) to move focus between buttons
- `enter` / `space` to select the focused button
- `y` / `n` as direct shortcuts (underlined in the button labels to hint this)
- `esc` always cancels

The focused button is highlighted in pink, the other in dark gray. If there are active downloads, the modal also shows how many will be paused.

## Changes

- `model.go` — new `QuitConfirmState` and `quitConfirmFocused` field
- `keys.go` — new `QuitConfirmKeyMap` with navigation/select/cancel bindings
- `update.go` — quit key now enters `QuitConfirmState`; new case handles navigation and confirmation
- `view.go` — `viewQuitConfirm()` renders the modal with styled buttons
- `update_test.go` — updated `TestUpdate_QuitCancelsEnqueueContext` for the two-step flow

## Testing

go test ./...

@SuperCoolPencil requesting review when you get a chance

<h3>Greptile Summary</h3>

This PR adds a two-step quit confirmation modal that intercepts `ctrl+c`/`ctrl+q` from `DashboardState`, presenting a focused "Yep!" / "Nope" button pair before committing to shutdown. Previous review-cycle feedback has been thoroughly addressed: `y`/`n` shortcuts now use the `key.Binding` system, the `confirmQuit`/`cancelQuit` helpers eliminate duplicated shutdown logic, `ctrl+c`/`ctrl+q` are included in the `Cancel` binding so the modal never swallows those keys, and the `renderBtn` helper consolidates button rendering. The `shuttingDown` flag being checked before `QuitConfirmState` in `View()` means the "Shutting Down" progress modal correctly supersedes the confirmation modal once the user confirms — no explicit state reset is needed in `confirmQuit`.

Key changes:
- `model.go` — new `QuitConfirmState` (14) and `quitConfirmFocused int` field
- `keys.go` — `QuitConfirmKeyMap` with `Left`, `Right`, `Yes`, `No`, `Select`, and `Cancel` (includes `ctrl+c`/`ctrl+q`) bindings
- `update.go` — quit key transitions to `QuitConfirmState`; dedicated case handles all navigation and confirmation paths via extracted helpers
- `view.go` — `viewQuitConfirm()` renders an interactive 60×10 modal with underlined button-label hints and a live active-download count
- Tests — comprehensive coverage across navigation aliases, shortcuts, cancellation paths, and tiny-terminal rendering

<h3>Confidence Score: 4/5</h3>

- Safe to merge with minor test coverage gap for uppercase Y/N shortcut variants.
- The implementation is clean and well-tested, all significant previous review feedback has been incorporated, and the state machine is correct (shuttingDown guard in View() prevents modal flicker on confirmation). The only remaining gaps are missing tests for the uppercase `Y`/`N` binding variants and a comment that describes values rather than design intent — neither is a blocking concern.
- internal/tui/update_test.go — uppercase Y/N shortcut variants have no test coverage despite being explicitly listed in the key bindings.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| internal/tui/keys.go | Adds `QuitConfirmKeyMap` with `Left`, `Right`, `Yes`, `No`, `Select`, and `Cancel` bindings (including `ctrl+c`/`ctrl+q` on Cancel); `ShortHelp`/`FullHelp` expose only `Select` and `Cancel`, omitting `Yes`/`No` from the help bar (acknowledged in prior review cycle). |
| internal/tui/model.go | Adds `QuitConfirmState` constant (14) and `quitConfirmFocused int` field; comment on the field maps values to button labels but doesn't explain the design choice of `int` over `bool`. |
| internal/tui/update.go | Quit key in `DashboardState` now sets `QuitConfirmState` instead of shutting down immediately; new `QuitConfirmState` case cleanly handles navigation, `Yes`/`No` shortcuts, `Select`, and `Cancel` (including `ctrl+c`/`ctrl+q`) via proper `key.Matches` calls with extracted `confirmQuit`/`cancelQuit` helpers. |
| internal/tui/update_test.go | Comprehensive test suite covering arrow/vim/tab navigation, `esc`/`ctrl+c`/`ctrl+q` cancellation, `y`/`n` shortcuts, `space` confirm, and unrelated key no-ops; missing tests for uppercase `Y`/`N` variants that are explicitly part of the bindings. |
| internal/tui/view.go | New `viewQuitConfirm` renders an interactive 60×10 modal with underlined-letter buttons, active-download count, and a help bar; `shuttingDown` is checked before `QuitConfirmState` in `View()` so the "Shutting Down" modal correctly supersedes the quit-confirm modal on confirmation, making the state reset a non-issue. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A([DashboardState]) -->|ctrl+c / ctrl+q| B[QuitConfirmState\nfocused = 0 ← Yep!]
    B -->|← / h / → / l / tab| C{Toggle focused\n1 - quitConfirmFocused}
    C --> B
    B -->|y / Y\nor enter/space on Yep!| D[confirmQuit\ncancelEnqueue\nshuttingDown = true]
    B -->|n / N\nor enter/space on Nope\nor esc / ctrl+c / ctrl+q| E[cancelQuit\nstate = DashboardState\nfocused = 0]
    D --> F[shutdownCmd]
    F --> G([Shutting Down modal\nView checks shuttingDown first])
    G -->|shutdownCompleteMsg| H([tea.Quit])
    E --> A
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `internal/tui/update_test.go`, line 598-631 ([link](https://github.com/surge-downloader/surge/blob/bde52533c3502c15d9466241857b02244ea3e97d/internal/tui/update_test.go#L598-L631)) 

   <a href="#">P2</a> **Missing edge-case test coverage for the confirmation modal**

   The updated test only exercises the "Yes" path (Enter with default focus). The following scenarios are currently untested:

   - **Cancel via `n`/`N` key** — modal should close, `shuttingDown` should remain `false`, and `quitConfirmFocused` should reset to `0`.
   - **Cancel via `Esc`** — same expectations as above.
   - **Navigate to "No" then confirm via `Enter`** — `Right` arrow → Enter should cancel, not quit.
   - **`y` shortcut** — should trigger shutdown immediately without the Select path.
   - **`cancelEnqueue` is `nil`** — confirming should not panic when no enqueue context is active.

   These paths collectively guard the two-step flow against regressions at every branch.

   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   <details><summary>Prompt To Fix With AI</summary>
   
   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_test.go
   Line: 598-631
   
   Comment:
   **Missing edge-case test coverage for the confirmation modal**
   
   The updated test only exercises the "Yes" path (Enter with default focus). The following scenarios are currently untested:
   
   - **Cancel via `n`/`N` key** — modal should close, `shuttingDown` should remain `false`, and `quitConfirmFocused` should reset to `0`.
   - **Cancel via `Esc`** — same expectations as above.
   - **Navigate to "No" then confirm via `Enter`** — `Right` arrow → Enter should cancel, not quit.
   - **`y` shortcut** — should trigger shutdown immediately without the Select path.
   - **`cancelEnqueue` is `nil`** — confirming should not panic when no enqueue context is active.
   
   These paths collectively guard the two-step flow against regressions at every branch.
   
   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: internal/tui/update_test.go
Line: 697-712

Comment:
**Uppercase `Y`/`N` shortcut paths untested**

`TestQuitConfirm_YShortcutConfirms` only exercises `Code: 'y'` and `TestQuitConfirm_NShortcutCancels` only exercises `Code: 'n'`. Both the `Yes` and `No` bindings explicitly include uppercase variants (`"Y"` and `"N"` respectively), so those code paths have no test coverage. Consider adding tests for the uppercase variants to lock in that the binding actually matches them:

```go
func TestQuitConfirm_UppercaseYShortcutConfirms(t *testing.T) {
    ctx, cancel := context.WithCancel(context.Background())
    m := newQuitConfirmModel()
    m.enqueueCtx = ctx
    m.cancelEnqueue = cancel

    updated, _ := m.Update(tea.KeyPressMsg{Code: 'Y'})
    m2 := updated.(RootModel)
    if !m2.shuttingDown {
        t.Fatal("expected Y to begin shutdown")
    }
    select {
    case <-ctx.Done():
    default:
        t.Fatal("expected Y to cancel enqueue context")
    }
}

func TestQuitConfirm_UppercaseNShortcutCancels(t *testing.T) {
    m := newQuitConfirmModel()
    updated, _ := m.Update(tea.KeyPressMsg{Code: 'N'})
    m2 := updated.(RootModel)
    if m2.state != DashboardState {
        t.Fatal("expected N to return to dashboard")
    }
}
```

**Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/model.go
Line: 161-162

Comment:
**Comment describes values, not rationale**

The comment documents what `0` and `1` map to, but doesn't explain why `int` is preferred over `bool` here — for instance, that using an integer index makes the toggle `1 - m.quitConfirmFocused` work uniformly if a third button is ever added, or simply that it mirrors the focus-index pattern used elsewhere in the model (`catMgrEditField`, `focusedInput`). A comment that surfaces the design intent would be more durable:

```suggestion
	// int index rather than bool so the toggle (1 - quitConfirmFocused) stays uniform
	// if a third button is ever added; 0 = Yep!, 1 = Nope
	quitConfirmFocused int
```

**Rule Used:** What: Comments must explain *why* code exists, not... ([source](https://app.greptile.com/review/custom-context?memory=4e45ef13-32c2-4753-8060-a44ef767144d))

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ["chore: go fmt"](https://github.com/surge-downloader/surge/commit/1bab9575a6769804f58c5d49cf94af9b2ae08d3f)</sub>

**Context used:**

- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
- Rule used - What: Comments must explain *why* code exists, not... ([source](https://app.greptile.com/review/custom-context?memory=4e45ef13-32c2-4753-8060-a44ef767144d))

### CODE REVIEW

**Comment by standard-librarian** [comment]
@SuperCoolPencil addressed the Greptile review comments and improved the coverage. Ready for your review whenever you get a chance!​​​​​​​​​​​​​​​​

**Comment by SuperCoolPencil** [comment]
Screenshot From 2026-03-21 11-15-31
Screenshot From 2026-03-21 11-15-27


Hey @standard-librarian 

I saw the changes, they look good - aligned with the repo.

A couple of concerns! 
- The current layout for the modal looks too cramped. A bit of padding would be nice
- I don't understand the need for arrow keys
- I strongly feel like the key hints should not take more than a single line (ie. all the key hints should be on one line)
- There is an existing modal in view.go:
```go
	if m.shuttingDown {
		modal := components.ConfirmationModal{
			Title:       "Shutting Down",
			Message:     "Pausing downloads and saving resume state...",
			Detail:      "Please wait",
			Keys:        components.ConfirmationKeyMap{},
			Help:        m.help,
			BorderColor: colors.NeonCyan,
			Width:       60,
			Height:      10,
		}
		box := modal.RenderWithBtopBox(renderBtopBox, PaneTitleStyle)
		return m.wrapView(m.renderModalWithOverlay(box))
	}
```
I feel like the confirmation modal should transition to the shutting down modal have the same dimensions so as to have consistency across the repo

---

Please let me know your thoughts on this. If you have any concerns, we can discuss them and work through them

**Comment by standard-librarian** [comment]
@SuperCoolPencil thank you for the feedback, I will be working on these points, do you think the navigation with the arrow keys should be removed ? I added them because they feel like a natural way or navigation when seeing the buttons, and added the y/n so it matches the design in https://github.com/surge-downloader/Surge/issues/273#issue-4109038628

**Comment by SuperCoolPencil** [comment]
I feel like you can have them, but there's no need for hinting?

Maybe I'm completely off here but they feel trivial.

**Comment by standard-librarian** [comment]
I get it, yeah this makes sense i will remove them and also to keep the repo consistent

**Comment by standard-librarian** [comment]
Sure, no worries, I will then remove pressing shift + tab to go back

File: internal/tui/keys.go [hunk 1]
```diff
@@ -438,6 +449,28 @@ var Keys = KeyMap{
 		Tab:    key.NewBinding(key.WithKeys("tab"), key.WithHelp("tab", "next field")),
 		Close:  key.NewBinding(key.WithKeys("esc"), key.WithHelp("esc", "save & close")),
 	},
+	QuitConfirm: QuitConfirmKeyMap{
+		Left: key.NewBinding(
+			key.WithKeys("left", "h"),
+		),
+		Right: key.NewBinding(
```
[the remark this answers is not in the document]
Comment (standard-librarian) [reply internal/tui/keys.go:456]: `shift+tab` is removed as we use `tab` circularly
